### PR TITLE
Optimize DATA reads with a cross-line scan adapted from uponusolutions

### PR DESCRIPTION
**Pull request by Jabberwocky238**
Large SMTP DATA bodies currently pass every byte through `ReadByte` and the dot/CRLF state machine. This patch copies buffered spans across ordinary CRLFs, returning to the existing state machine around potential dot-prefixed lines.

The cross-line scanning approach is adapted from **ml1nk / UPONU Solutions** in [uponusolutions/go-smtp, `internal/textsmtp/dotreader.go` at `86ff2622fb52`](https://github.com/uponusolutions/go-smtp/blob/86ff2622fb52f86371265b74a976333ff53c10a0/internal/textsmtp/dotreader.go). Source attribution and the MIT copyright notice are included in `data.go`.

Only **`data.go` changes in production**; the other two files are tests. The patch integrates the scanning approach into this library's existing reader. It introduces no dependency, package move, public API, buffer-size option, or line-limit change. The normal 4 KiB connection buffer is retained.

The original state machine handles the first line, dot unstuffing, the terminator, incomplete input, message limits, and fragmented input/output. The fast path preserves its handling of consecutive CRs too. In the contributor's unmodified reader we reproduced two boundary failures: an initial `..first` was not unstuffed, and an empty `.\r\n` blocked on a live connection waiting for the SMTP response (a finite reader returned the terminator with `io.ErrUnexpectedEOF`). The adapted version covers both cases. It only inspects already-buffered spans and never waits for five bytes beyond a possible terminator.

`Peek`/`Discard` keeps each span bounded by the caller's output buffer. `ReadSlice` can consume a larger span, and `UnreadByte` can restore only one byte; supporting that alternative would require additional remainder handling.

### Validation

- `go test -race ./...`
- `go test -run '^$' -fuzz '^FuzzDataReader$' -fuzztime=20s -parallel=4`
- 5,000 deterministic differential cases against the pre-optimization reader, plus fragmented input/output, first-line dots, a live empty message, zero-length reads, size limits, and preservation of a following SMTP command.
- Embedding mail service: `make test` passed, including race checks, native S3-backed SMTP/POP3/IMAP/JMAP integration, deployment/install tests, and a 32 MiB attachment round trip.

### Reader benchmark

Apple M4, macOS arm64, Go 1.25.14; the same 2 GiB MIME-like streaming fixture, default 4 KiB reader and 128 KiB copy buffer, three single-iteration runs, baseline and revised implementation run sequentially without profiling/race instrumentation:

```sh
GOTOOLCHAIN=go1.25.14 go test -run '^$' -bench '^BenchmarkDataReader/MIME$' -benchtime=1x -count=3
```

| Version | Runs (ms) | Median (ms) | B/op | allocs/op |
| --- | --- | ---: | ---: | ---: |
| Previous PR head `f3ad4805e305` | 634.885, 627.195, 648.223 | 634.885 | 4,360 | 7 |
| Cross-line adaptation `b0673510e580` | 427.729, 435.903, 431.226 | 431.226 | 4,360 | 7 |

The revised reader takes **32.1% less time** (about **1.47× throughput**) in this local comparison. This measures the DATA reader, not line-length validation, network/TLS, MIME processing, or storage. The earlier application's 8.583 s upload result used the removed buffer/line-limit changes and does not describe this revision.

A fresh end-to-end run after pinning the published revision passed the 2 GiB attachment round trip and all download hashes, with 167.4 MiB maximum sampled RSS (256 MiB ceiling). SMTP upload took 10.099 s. This is slower than the historical 8.583 s result with the broader buffer/line-limit changes; those unpaired runs have different settings, so this revision does not claim an end-to-end improvement.

### CODE REVIEW

File: data.go [hunk 1]
```diff
@@ -92,6 +93,23 @@ func (r *dataReader) Read(b []byte) (n int, err error) {
...
+			if available > len(b)-n {
+				available = len(b) - n
+			}
+			p, _ := r.r.Peek(available)
+			if i := bytes.IndexByte(p, '\r'); i >= 0 {
+				p = p[:i+1]
+				r.state = stateCR
+			}
+			n += copy(b[n:], p)
+			r.r.Discard(len(p))
```
Comment (emersion) [inline data.go:133]: Could we use `ReadSlice` + `UnreadByte` instead?

**Comment by ml1nk** [comment]
Some time ago I rewrote dot & bdat reader/writer, it would be nice to compare your changes against these
https://github.com/uponusolutions/go-smtp/blob/main/internal/textsmtp/dotreader.go
to see how it compares. I cannot guarantee that it works in all cases, so I never tried to get it back upstream, but at least it works with the test cases there are.

**Comment by Jabberwocky238** [comment]
Thanks @ml1nk for sharing the implementation, and @emersion for the review. I have updated this PR in b0673510e580 to adapt the cross-line scan from uponusolutions/go-smtp, with source attribution and MIT credit. The production diff is now confined to `data.go`; the buffer-size API and line-limit changes have been removed.

In a sequential comparison against the previous PR head, the same 2 GiB reader benchmark improved from a 634.885 ms median to 431.226 ms (32.1% less time), with unchanged allocations. Details and scope are in the updated description.

I reproduced two edge cases in the original fork: the first line's `..first` was not unstuffed, and an empty DATA terminator blocked when the peer kept the connection open waiting for a reply. This adaptation keeps our existing state machine around those boundaries and adds regression and differential tests. It also preserves the old handling of consecutive CRs.

I retained bounded `Peek`/`Discard` because `ReadSlice` may consume more than the caller's output buffer can hold, while `UnreadByte` restores only one byte. Avoiding additional pending-slice state keeps this patch small.

Library race/fuzz tests, the embedding application's full suite, and a 2 GiB streamed attachment round trip passed. The new application's SMTP upload was 10.099 s; the reader improvement should not be read as a 32% end-to-end speedup, especially after removing the other optimizations.

**Comment by Jabberwocky238** [comment]
Adapted uponusolutions' reader in one file to keep pace with upstream while preserving our API.

**Comment by ml1nk** [comment]
... and I used your testcases to fix up my dotreader :)

**Comment by ml1nk** [comment]
@Jabberwocky238 
I reworked the benchmarks to see the performance differences in the implementations and perf-flags like Pipelining. I currently working on adding bdat reader/writer to the benchmarks. If you have time, you can change the upstream implementation (copy from this repo inside a testcase, https://github.com/uponusolutions/go-smtp/blob/main/internal/textsmtp/dotreader_upstream_test.go) and test it on you machine how it performs. Your change and mine should have roughly the same perf.

~~~
goos: linux
goarch: amd64
pkg: github.com/uponusolutions/go-smtp
cpu: AMD Ryzen AI 9 HX 370 w/ Radeon 890M
                                                             │ .bench/20260913T012223.txt │
                                                             │           sec/op           │
Mailer/Small/NoChunking/NoPipelining/CloseConn/MinimalReader                  102.7µ ± 1%
Mailer/Small/NoChunking/NoPipelining/CloseConn/BufferReader                   100.7µ ± 0%
Mailer/Small/NoChunking/NoPipelining/ReuseConn/MinimalReader                  43.93µ ± 0%
Mailer/Small/NoChunking/NoPipelining/ReuseConn/BufferReader                   42.69µ ± 1%
Mailer/Small/NoChunking/Pipelining/CloseConn/MinimalReader                    83.23µ ± 1%
Mailer/Small/NoChunking/Pipelining/CloseConn/BufferReader                     81.77µ ± 1%
Mailer/Small/NoChunking/Pipelining/ReuseConn/MinimalReader                    25.05µ ± 0%
Mailer/Small/NoChunking/Pipelining/ReuseConn/BufferReader                     23.80µ ± 1%
Mailer/Small/Chunking/NoPipelining/CloseConn/MinimalReader                    97.26µ ± 1%
Mailer/Small/Chunking/NoPipelining/CloseConn/BufferReader                     92.82µ ± 0%
Mailer/Small/Chunking/NoPipelining/ReuseConn/MinimalReader                    34.68µ ± 0%
Mailer/Small/Chunking/NoPipelining/ReuseConn/BufferReader                     34.77µ ± 0%
Mailer/Small/Chunking/Pipelining/CloseConn/MinimalReader                      84.87µ ± 1%
Mailer/Small/Chunking/Pipelining/CloseConn/BufferReader                       80.64µ ± 0%
Mailer/Small/Chunking/Pipelining/ReuseConn/MinimalReader                      22.15µ ± 0%
Mailer/Small/Chunking/Pipelining/ReuseConn/BufferReader                       22.28µ ± 0%
Mailer/Large/NoChunking/NoPipelining/CloseConn/MinimalReader                  13.79m ± 0%
Mailer/Large/NoChunking/NoPipelining/CloseConn/BufferReader                   13.80m ± 1%
Mailer/Large/NoChunking/NoPipelining/ReuseConn/MinimalReader                  21.07m ± 1%
Mailer/Large/NoChunking/NoPipelining/ReuseConn/BufferReader                   21.10m ± 1%
Mailer/Large/NoChunking/Pipelining/CloseConn/MinimalReader                    13.83m ± 0%
Mailer/Large/NoChunking/Pipelining/CloseConn/BufferReader                     13.77m ± 1%
Mailer/Large/NoChunking/Pipelining/ReuseConn/MinimalReader                    21.12m ± 1%
Mailer/Large/NoChunking/Pipelining/ReuseConn/BufferReader                     21.09m ± 0%
Mailer/Large/Chunking/NoPipelining/CloseConn/MinimalReader                    10.21m ± 1%
Mailer/Large/Chunking/NoPipelining/CloseConn/BufferReader                     9.146m ± 0%
Mailer/Large/Chunking/NoPipelining/ReuseConn/MinimalReader                    9.433m ± 1%
Mailer/Large/Chunking/NoPipelining/ReuseConn/BufferReader                     8.844m ± 1%
Mailer/Large/Chunking/Pipelining/CloseConn/MinimalReader                      10.10m ± 1%
Mailer/Large/Chunking/Pipelining/CloseConn/BufferReader                       9.121m ± 1%
Mailer/Large/Chunking/Pipelining/ReuseConn/MinimalReader                      9.547m ± 1%
Mailer/Large/Chunking/Pipelining/ReuseConn/BufferReader                       8.894m ± 1%
geomean                                                                       811.8µ

                                                             │ .bench/20260913T012223.txt │
                                                             │            B/s             │
Mailer/Small/NoChunking/NoPipelining/CloseConn/MinimalReader                 32.83Mi ± 1%
Mailer/Small/NoChunking/NoPipelining/CloseConn/BufferReader                  33.47Mi ± 0%
Mailer/Small/NoChunking/NoPipelining/ReuseConn/MinimalReader                 76.71Mi ± 0%
Mailer/Small/NoChunking/NoPipelining/ReuseConn/BufferReader                  78.94Mi ± 1%
Mailer/Small/NoChunking/Pipelining/CloseConn/MinimalReader                   40.49Mi ± 1%
Mailer/Small/NoChunking/Pipelining/CloseConn/BufferReader                    41.22Mi ± 1%
Mailer/Small/NoChunking/Pipelining/ReuseConn/MinimalReader                   134.6Mi ± 0%
Mailer/Small/NoChunking/Pipelining/ReuseConn/BufferReader                    141.6Mi ± 1%
Mailer/Small/Chunking/NoPipelining/CloseConn/MinimalReader                   34.65Mi ± 1%
Mailer/Small/Chunking/NoPipelining/CloseConn/BufferReader                    36.31Mi ± 0%
Mailer/Small/Chunking/NoPipelining/ReuseConn/MinimalReader                   97.19Mi ± 0%
Mailer/Small/Chunking/NoPipelining/ReuseConn/BufferReader                    96.93Mi ± 0%
Mailer/Small/Chunking/Pipelining/CloseConn/MinimalReader                     39.71Mi ± 1%
Mailer/Small/Chunking/Pipelining/CloseConn/BufferReader                      41.79Mi ± 0%
Mailer/Small/Chunking/Pipelining/ReuseConn/MinimalReader                     152.2Mi ± 0%
Mailer/Small/Chunking/Pipelining/ReuseConn/BufferReader                      151.3Mi ± 0%
Mailer/Large/NoChunking/NoPipelining/CloseConn/MinimalReader                 925.4Mi ± 0%
Mailer/Large/NoChunking/NoPipelining/CloseConn/BufferReader                  924.2Mi ± 1%
Mailer/Large/NoChunking/NoPipelining/ReuseConn/MinimalReader                 605.3Mi ± 1%
Mailer/Large/NoChunking/NoPipelining/ReuseConn/BufferReader                  604.5Mi ± 1%
Mailer/Large/NoChunking/Pipelining/CloseConn/MinimalReader                   922.5Mi ± 0%
Mailer/Large/NoChunking/Pipelining/CloseConn/BufferReader                    926.5Mi ± 1%
Mailer/Large/NoChunking/Pipelining/ReuseConn/MinimalReader                   604.1Mi ± 1%
Mailer/Large/NoChunking/Pipelining/ReuseConn/BufferReader                    604.8Mi ± 0%
Mailer/Large/Chunking/NoPipelining/CloseConn/MinimalReader                   1.221Gi ± 1%
Mailer/Large/Chunking/NoPipelining/CloseConn/BufferReader                    1.362Gi ± 0%
Mailer/Large/Chunking/NoPipelining/ReuseConn/MinimalReader                   1.321Gi ± 1%
Mailer/Large/Chunking/NoPipelining/ReuseConn/BufferReader                    1.409Gi ± 1%
Mailer/Large/Chunking/Pipelining/CloseConn/MinimalReader                     1.233Gi ± 1%
Mailer/Large/Chunking/Pipelining/CloseConn/BufferReader                      1.366Gi ± 1%
Mailer/Large/Chunking/Pipelining/ReuseConn/MinimalReader                     1.305Gi ± 1%
Mailer/Large/Chunking/Pipelining/ReuseConn/BufferReader                      1.401Gi ± 1%
geomean                                                                      255.4Mi

                                                             │ .bench/20260913T012223.txt │
                                                             │            B/op            │
Mailer/Small/NoChunking/NoPipelining/CloseConn/MinimalReader                 70.13Ki ± 0%
Mailer/Small/NoChunking/NoPipelining/CloseConn/BufferReader                  38.24Ki ± 0%
Mailer/Small/NoChunking/NoPipelining/ReuseConn/MinimalReader                 49.23Ki ± 0%
Mailer/Small/NoChunking/NoPipelining/ReuseConn/BufferReader                  17.33Ki ± 0%
Mailer/Small/NoChunking/Pipelining/CloseConn/MinimalReader                   70.27Ki ± 0%
Mailer/Small/NoChunking/Pipelining/CloseConn/BufferReader                    38.38Ki ± 0%
Mailer/Small/NoChunking/Pipelining/ReuseConn/MinimalReader                   49.34Ki ± 0%
Mailer/Small/NoChunking/Pipelining/ReuseConn/BufferReader                    17.44Ki ± 0%
Mailer/Small/Chunking/NoPipelining/CloseConn/MinimalReader                   2.037Mi ± 0%
Mailer/Small/Chunking/NoPipelining/CloseConn/BufferReader                    38.16Ki ± 0%
Mailer/Small/Chunking/NoPipelining/ReuseConn/MinimalReader                   17.22Ki ± 0%
Mailer/Small/Chunking/NoPipelining/ReuseConn/BufferReader                    17.24Ki ± 0%
Mailer/Small/Chunking/Pipelining/CloseConn/MinimalReader                     2.037Mi ± 0%
Mailer/Small/Chunking/Pipelining/CloseConn/BufferReader                      38.27Ki ± 0%
Mailer/Small/Chunking/Pipelining/ReuseConn/MinimalReader                     17.29Ki ± 0%
Mailer/Small/Chunking/Pipelining/ReuseConn/BufferReader                      17.33Ki ± 0%
Mailer/Large/NoChunking/NoPipelining/CloseConn/MinimalReader                 31.37Mi ± 0%
Mailer/Large/NoChunking/NoPipelining/CloseConn/BufferReader                  31.33Mi ± 0%
Mailer/Large/NoChunking/NoPipelining/ReuseConn/MinimalReader                 31.34Mi ± 0%
Mailer/Large/NoChunking/NoPipelining/ReuseConn/BufferReader                  31.31Mi ± 0%
Mailer/Large/NoChunking/Pipelining/CloseConn/MinimalReader                   31.37Mi ± 0%
Mailer/Large/NoChunking/Pipelining/CloseConn/BufferReader                    31.33Mi ± 0%
Mailer/Large/NoChunking/Pipelining/ReuseConn/MinimalReader                   31.34Mi ± 0%
Mailer/Large/NoChunking/Pipelining/ReuseConn/BufferReader                    31.31Mi ± 0%
Mailer/Large/Chunking/NoPipelining/CloseConn/MinimalReader                   33.34Mi ± 0%
Mailer/Large/Chunking/NoPipelining/CloseConn/BufferReader                    31.34Mi ± 0%
Mailer/Large/Chunking/NoPipelining/ReuseConn/MinimalReader                   31.33Mi ± 0%
Mailer/Large/Chunking/NoPipelining/ReuseConn/BufferReader                    31.32Mi ± 0%
Mailer/Large/Chunking/Pipelining/CloseConn/MinimalReader                     33.34Mi ± 0%
Mailer/Large/Chunking/Pipelining/CloseConn/BufferReader                      31.34Mi ± 0%
Mailer/Large/Chunking/Pipelining/ReuseConn/MinimalReader                     31.33Mi ± 0%
Mailer/Large/Chunking/Pipelining/ReuseConn/BufferReader                      31.32Mi ± 0%
geomean                                                                      1.268Mi

                                                             │ .bench/20260913T012223.txt │
                                                             │         allocs/op          │
Mailer/Small/NoChunking/NoPipelining/CloseConn/MinimalReader                   215.0 ± 0%
Mailer/Small/NoChunking/NoPipelining/CloseConn/BufferReader                    216.0 ± 0%
Mailer/Small/NoChunking/NoPipelining/ReuseConn/MinimalReader                   110.0 ± 0%
Mailer/Small/NoChunking/NoPipelining/ReuseConn/BufferReader                    111.0 ± 0%
Mailer/Small/NoChunking/Pipelining/CloseConn/MinimalReader                     220.0 ± 0%
Mailer/Small/NoChunking/Pipelining/CloseConn/BufferReader                      221.0 ± 0%
Mailer/Small/NoChunking/Pipelining/ReuseConn/MinimalReader                     114.0 ± 0%
Mailer/Small/NoChunking/Pipelining/ReuseConn/BufferReader                      115.0 ± 0%
Mailer/Small/Chunking/NoPipelining/CloseConn/MinimalReader                     215.0 ± 0%
Mailer/Small/Chunking/NoPipelining/CloseConn/BufferReader                      216.0 ± 0%
Mailer/Small/Chunking/NoPipelining/ReuseConn/MinimalReader                     107.0 ± 0%
Mailer/Small/Chunking/NoPipelining/ReuseConn/BufferReader                      109.0 ± 0%
Mailer/Small/Chunking/Pipelining/CloseConn/MinimalReader                       219.0 ± 0%
Mailer/Small/Chunking/Pipelining/CloseConn/BufferReader                        220.0 ± 0%
Mailer/Small/Chunking/Pipelining/ReuseConn/MinimalReader                       110.0 ± 0%
Mailer/Small/Chunking/Pipelining/ReuseConn/BufferReader                        112.0 ± 0%
Mailer/Large/NoChunking/NoPipelining/CloseConn/MinimalReader                   238.0 ± 0%
Mailer/Large/NoChunking/NoPipelining/CloseConn/BufferReader                    239.0 ± 0%
Mailer/Large/NoChunking/NoPipelining/ReuseConn/MinimalReader                   133.0 ± 0%
Mailer/Large/NoChunking/NoPipelining/ReuseConn/BufferReader                    134.0 ± 0%
Mailer/Large/NoChunking/Pipelining/CloseConn/MinimalReader                     243.0 ± 0%
Mailer/Large/NoChunking/Pipelining/CloseConn/BufferReader                      244.0 ± 0%
Mailer/Large/NoChunking/Pipelining/ReuseConn/MinimalReader                     137.0 ± 0%
Mailer/Large/NoChunking/Pipelining/ReuseConn/BufferReader                      138.0 ± 0%
Mailer/Large/Chunking/NoPipelining/CloseConn/MinimalReader                     334.0 ± 0%
Mailer/Large/Chunking/NoPipelining/CloseConn/BufferReader                      335.0 ± 0%
Mailer/Large/Chunking/NoPipelining/ReuseConn/MinimalReader                     226.0 ± 0%
Mailer/Large/Chunking/NoPipelining/ReuseConn/BufferReader                      228.0 ± 0%
Mailer/Large/Chunking/Pipelining/CloseConn/MinimalReader                       338.0 ± 0%
Mailer/Large/Chunking/Pipelining/CloseConn/BufferReader                        339.0 ± 0%
Mailer/Large/Chunking/Pipelining/ReuseConn/MinimalReader                       229.0 ± 0%
Mailer/Large/Chunking/Pipelining/ReuseConn/BufferReader                        231.0 ± 0%
geomean                                                                        186.5

pkg: github.com/uponusolutions/go-smtp/internal/textsmtp
                                        │ .bench/20260913T012223.txt │
                                        │           sec/op           │
Dot/Read/Binary/MinimalReader/Upstream                   10.53m ± 8%
Dot/Read/Binary/MinimalReader/Fork                       719.8µ ± 0%
Dot/Read/Binary/BufferReader/Upstream                    11.32m ± 9%
Dot/Read/Binary/BufferReader/Fork                        722.9µ ± 1%
Dot/Read/Text/MinimalReader/Upstream                     9.813m ± 8%
Dot/Read/Text/MinimalReader/Fork                         488.3µ ± 1%
Dot/Read/Text/BufferReader/Upstream                      10.37m ± 5%
Dot/Read/Text/BufferReader/Fork                          485.5µ ± 1%
Dot/Write/Binary/MinimalReader/Upstream                  9.240m ± 2%
Dot/Write/Binary/MinimalReader/Fork                      396.2µ ± 0%
Dot/Write/Binary/BufferReader/Upstream                   8.546m ± 4%
Dot/Write/Binary/BufferReader/Fork                       459.7µ ± 0%
Dot/Write/Text/MinimalReader/Upstream                    8.725m ± 0%
Dot/Write/Text/MinimalReader/Fork                        403.0µ ± 2%
Dot/Write/Text/BufferReader/Upstream                     8.857m ± 1%
Dot/Write/Text/BufferReader/Fork                         473.6µ ± 1%
geomean                                                  2.207m

                                        │ .bench/20260913T012223.txt │
                                        │            B/s             │
Dot/Read/Binary/MinimalReader/Upstream                 381.4Mi ±  7%
Dot/Read/Binary/MinimalReader/Fork                     5.448Gi ±  0%
Dot/Read/Binary/BufferReader/Upstream                  355.1Mi ± 10%
Dot/Read/Binary/BufferReader/Fork                      5.424Gi ±  1%
Dot/Read/Text/MinimalReader/Upstream                   408.3Mi ±  7%
Dot/Read/Text/MinimalReader/Fork                       8.013Gi ±  1%
Dot/Read/Text/BufferReader/Upstream                    386.4Mi ±  6%
Dot/Read/Text/BufferReader/Fork                        8.059Gi ±  1%
Dot/Write/Binary/MinimalReader/Upstream                432.9Mi ±  2%
Dot/Write/Binary/MinimalReader/Fork                    9.858Gi ±  0%
Dot/Write/Binary/BufferReader/Upstream                 468.1Mi ±  4%
Dot/Write/Binary/BufferReader/Fork                     8.497Gi ±  0%
Dot/Write/Text/MinimalReader/Upstream                  458.5Mi ±  0%
Dot/Write/Text/MinimalReader/Fork                      9.694Gi ±  2%
Dot/Write/Text/BufferReader/Upstream                   451.6Mi ±  1%
Dot/Write/Text/BufferReader/Fork                       8.248Gi ±  1%
geomean                                                1.772Gi

                                        │ .bench/20260913T012223.txt │
                                        │            B/op            │
Dot/Read/Binary/MinimalReader/Upstream                  4.253Ki ± 0%
Dot/Read/Binary/MinimalReader/Fork                      4.177Ki ± 0%
Dot/Read/Binary/BufferReader/Upstream                   4.237Ki ± 0%
Dot/Read/Binary/BufferReader/Fork                       4.161Ki ± 0%
Dot/Read/Text/MinimalReader/Upstream                    4.239Ki ± 0%
Dot/Read/Text/MinimalReader/Fork                        4.175Ki ± 0%
Dot/Read/Text/BufferReader/Upstream                     4.237Ki ± 0%
Dot/Read/Text/BufferReader/Fork                         4.159Ki ± 0%
Dot/Write/Binary/MinimalReader/Upstream                 4.141Ki ± 0%
Dot/Write/Binary/MinimalReader/Fork                     4.125Ki ± 0%
Dot/Write/Binary/BufferReader/Upstream                  36.12Ki ± 0%
Dot/Write/Binary/BufferReader/Fork                      36.11Ki ± 0%
Dot/Write/Text/MinimalReader/Upstream                   4.141Ki ± 0%
Dot/Write/Text/MinimalReader/Fork                       4.125Ki ± 0%
Dot/Write/Text/BufferReader/Upstream                    36.12Ki ± 0%
Dot/Write/Text/BufferReader/Fork                        36.11Ki ± 0%
geomean                                                 7.167Ki

                                        │ .bench/20260913T012223.txt │
                                        │         allocs/op          │
Dot/Read/Binary/MinimalReader/Upstream                    4.000 ± 0%
Dot/Read/Binary/MinimalReader/Fork                        4.000 ± 0%
Dot/Read/Binary/BufferReader/Upstream                     4.000 ± 0%
Dot/Read/Binary/BufferReader/Fork                         4.000 ± 0%
Dot/Read/Text/MinimalReader/Upstream                      4.000 ± 0%
Dot/Read/Text/MinimalReader/Fork                          4.000 ± 0%
Dot/Read/Text/BufferReader/Upstream                       4.000 ± 0%
Dot/Read/Text/BufferReader/Fork                           4.000 ± 0%
Dot/Write/Binary/MinimalReader/Upstream                   5.000 ± 0%
Dot/Write/Binary/MinimalReader/Fork                       4.000 ± 0%
Dot/Write/Binary/BufferReader/Upstream                    6.000 ± 0%
Dot/Write/Binary/BufferReader/Fork                        5.000 ± 0%
Dot/Write/Text/MinimalReader/Upstream                     5.000 ± 0%
Dot/Write/Text/MinimalReader/Fork                         4.000 ± 0%
Dot/Write/Text/BufferReader/Upstream                      6.000 ± 0%
Dot/Write/Text/BufferReader/Fork                          5.000 ± 0%
geomean                                                   4.449
~~~

**Comment by Jabberwocky238** [comment]
@ml1nk Thanks, I tested your updated reader against this PR on an Apple M4 (Go 1.25.3), using the same 4 KiB input buffer.

For a 2 GiB MIME stream, six alternating runs gave these median times:

| | This PR | Your reader |
|---|---:|---:|
| Reader only | 329 ms | 310 ms |
| With the existing line-length check | 1574 ms | 1557 ms |

Your reader was ~6% faster in isolation; with line checks the difference was ~1%, with overlapping timings. This wasn't an end-to-end SMTP test.

I also ran this PR's compatibility tests against your reader. Some still differ from upstream: consecutive CRs affect terminator detection, and truncated input with a small output buffer can return an error before delivering all buffered bytes. For example, reading truncated `hello\r\n` one byte at a time returned only `hell` before `io.ErrUnexpectedEOF`.

So there's a measurable reader improvement, but I'd want to preserve those edge cases before adopting the rest.

**Comment by ml1nk** [comment]
Thank you, I didn't checked the fragmented tests until now good catch. I did some changes to my dotreader again to fix these deviations, but one seems ok for me: I never remove loose \r  or \n. The upstream dotreader removes \r if it directly comes after \r\n. or . in the first line.
~~~
		{"dot-cr", ".\rx\r\n.\ry\r\n.\r\n", "\rx\r\n\ry\r\n", nil}, // original "x\r\n" -- why should \r be removed?
~~~
